pkg/trace/obfuscate/sql.go

```diff
@@ -159,14 +160,42 @@ func obfuscateSQLString(in string) (string, error) {
 		err       error
 		lastToken TokenKind
 	)
+
 	// call Scan() function until tokens are available or if a LEX_ERROR is raised. After
 	// retrieving a token, send it to the tokenFilter chains so that the token is discarded
 	// or replaced.
-	token, buff := tokenizer.Scan()
-	for ; token != EOFChar; token, buff = tokenizer.Scan() {
+	for {
+		pos := tokenizer.pos
+		hasEscaped := tokenizer.backslashQuote.hasEscaped
+		token, buff := tokenizer.Scan()
+		if token == EOFChar {
+			break
+		}
+
+		if hasEscaped != tokenizer.backslashQuote.hasEscaped {
+			// tokenizer has encountered its first escaped quote
+			tokenizer.backslashQuote.firstEscapedPos = pos
+			tokenizer.backslashQuote.firstEscapedOutputLen = out.Len()
+			tokenizer.backslashQuote.firstEscapedLastToken = lastToken
+		}
+
 		if token == LexError {
```

Comment:
This isn't the right solution. This approach means that the tokenizer can return errors for valid strings and that the bug isn't fixed, instead we are just working around it by inserting state into the tokenizer from an external component. The fix for this should simply be contained inside `scanString`, which should be able to correctly scan a string, regardless of escape characters.

We want tokenizer to be a standalone component which works, and not assume that some returned error might be because of a bug we know and treat it by hacking into the tokenizer. I would prefer that we don't fix this bug if it turns out to be too complicated...

Reply:
+1

Reply:
I disagree. See comment below. I don't believe it is possible for `scanString` to have all of the context possible to scan a string "regardless of escape characters". It is not possible for it to know whether a character should be an escape character or not.

Reply:
Right. I must apologise first because having looked at this too briefly last night around 10pm, I didn't give it enough attention to notice all the details. Now that you have simplified the variable names and I have read through it more carefully, including your thorough PR description, I think I get it.

The problem is that we are handling multiple types of SQL engines which may or may not escape quotes in different ways: some use `\'`, others use `''`, while some treat escape characters literally. When we first start parsing a query, we have to start with one of the assumptions from the git-go and can not change it mid-way. Thus, if we get it wrong, we need to start over with a new assumption. Do I have the correct understanding of the problem now?

If yes, I'd like to suggest a simplification, for the sake of readability. What do you think of this proposal:

* Have a config setting in the tokenizer which tells it whether to ignore escapes or not, as you already do now (`ignoreEscape`).
* Remove the `Seek` logic completely. If we fail, simply start over from scratch with the setting enabled.
* Add a method `HadEscape` on the tokenizer which reports whether an escape was found (just like you have `hasEscaped` now)
* Store an atomic state (possibly at package level?) which tells us whether we should parse using `ignoreEscape` enabled by default. This way, we might get it wrong the first time, but on subsequent parses we will get it right every time. I think it's fine to consider that most of the time a customer running the trace-agent has a consistent SQL engine sending to it. This should remove the complexity of seeking and potentially keep performance virtually the same.

What are your thoughts? Could it work? I'm trying to remove as much complexity as possible. I want the tokenizer to be as easy to understand as possible to future readers.